dhall/src/Dhall/Core.hs

```diff
@@ -324,6 +325,37 @@ instance Pretty Import where
             Location -> " as Location"
             Code     -> ""
 
+-- | This wrapper around 'Prelude.Double' exists for its 'Eq' instance which is
+-- defined via the binary encoding of Dhall @Double@s.
+newtype DhallDouble = DhallDouble { getDhallDouble :: Double }
+    deriving (Show, Data, NFData, Generic)
+
+-- | This instance satisfies all the customary 'Eq' laws except substitutivity.
+--
+-- In particular:
+--
+-- >>> nan = DhallDouble (0/0)
+-- >>> nan == nan
+-- True
+--
+-- This instance is also consistent with with the binary encoding of Dhall @Double@s:
+--
+-- >>> toBytes n = Codec.Serialise.serialise (Dhall.Binary.encode (n :: DhallDouble))
+--
+-- prop> \a b -> (a == b) == (toBytes a == toBytes b)
+instance Eq DhallDouble where
+    DhallDouble a == DhallDouble b
+        | isNaN a && isNaN b = True
+        | otherwise          = a == b
```

Comment:
This instance, and using it in `conv`, makes `-0.0` and `0.0` equivalent.

I've opened https://github.com/dhall-lang/dhall-lang/issues/752 to find out whether we actually want this.

Reply:
I have updated the instance to treat `-0.0` and `0.0` as different, just as before this patch. 

Depending on the outcome of https://github.com/dhall-lang/dhall-lang/issues/752, we can still change this later.